### PR TITLE
NN-4455 NN-4454 refactor and updated paragraph descriptions

### PR DESCRIPTION
**Pull request by timoj58**
Rationale behind the refactoring

After review of the code, it became apparent that over various iterations of the feature previously the rules and logic has changed.  

Since the introduction of the youth offender paragraphs, various descriptions are shared, and the UI will now look up the codes via the offenceCodeLookupService, and not the column stored in the reportedOffence table.

Additionally, the code was built to support multiple nomis codes (ie the varags argument) and to group offenceCodes by offenceCode, however these were always unique.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/OffenceCodeLookupService.kt [hunk 1]
```diff
@@ -4,103 +4,46 @@ import org.springframework.stereotype.Service
...
-    allCodeBuilders.add(codeLookupBuilder)
-    return codeLookupBuilder
-  }
-
-  fun build(): Map<Int, OffenceCodeDetails> {
-    return allCodeBuilders.map { it.build() }.groupBy { it.offenceCode }
-      .mapValues {
-        if (it.value.size == 1) it.value[0] else throw RuntimeException("No single value provided for key ${it.key}")
+      true -> {
+        val res = youthOffenceDetailsByCode[offenceCode]
```
Comment (neilmendum) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/OffenceCodeLookupService.kt:28]: I think you could refactor this and the other branches' logic to
```
youthOffenceDetailsByCode[offenceCode] ?.let {
  OffenceCodeDetails(offenceCode = offenceCode, paragraphNumber = it.first, nomisCode = it.second, notCommittedOnOwnNomisCode = it.third)
}
```

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/OffenceCodeParagraphDescriptions.kt [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+  YOI_24_ADULT_21("Intentionally fails to work properly or, being required to work, refuses to do so"),
+  YOI_25_ADULT_22("Disobeys any lawful order"),
+  YOI_26_ADULT_23("Disobeys or fails to comply with any rule or regulation applying to them"),
+  DEFAULT("")
+}
+
+class OffenceCodeParagraphs {
+
+  private val lookup = mapOf(
+    "55:2" to "YOI_1_ADULT_1",
```
Comment (neilmendum) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/OffenceCodeParagraphDescriptions.kt:44]: We could make this a `Map<String, Descriptions>` for example `"55:2" to Descriptions.YOI_1_ADULT_1` which would simplify `getParagraphDescription`?

**Review by neilmendum — COMMENTED** [review]
Looks good! Made a few small comments

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/OffenceCodeLookupService.kt [hunk 3]
```diff
@@ -4,103 +4,46 @@ import org.springframework.stereotype.Service
...
-  val paragraph: OffenceCodeParagraph,
-)
-
-data class OffenceCodeParagraph(
   val paragraphNumber: String,
-  val paragraphDescription: String,
-  val paragraphCode: String = paragraphNumber,
+  val nomisCode: String,
+  val notCommittedOnOwnNomisCode: String,
+  val paragraphCode: String = paragraphNumber.replace("(", "").replace(")", ""),
```
[the remark this answers is not in the document]
Comment (timoj58) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/OffenceCodeLookupService.kt:43]: no, its a place holder for now as it seems a single code has a desc / code for nomis.  Mainly this was to highlight it.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/AdultOffenceCodes.kt [hunk 4]
```diff
@@ -2,147 +2,64 @@ package uk.gov.justice.digital.hmpps.hmppsmanageadjudicationsapi.services
 
 class AdultOffenceCodes {
   companion object {
-    // This information is taken from the spreadsheet of offences related to the ticket this work was done under.
+    // This information is taken from the spreadsheet of offences related to Triple(the ticket this work was done under.
```
[the remark this answers is not in the document]
Comment (timoj58) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/AdultOffenceCodes.kt:5]: yeah, i saw and will change!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/OffenceCodeParagraphDescriptions.kt (hunk 2 above)
Comment (timoj58) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/OffenceCodeParagraphDescriptions.kt:44]: Yeah good idea.  It was a string as i generated it that way, but makes sense now the test confirmed the mappings are correct.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/OffenceCodeLookupService.kt (hunk 1 above)
Comment (timoj58) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/OffenceCodeLookupService.kt:26]: updated

**Review by neilmendum — COMMENTED** [review]
LGTM

**Review by neilmendum — APPROVED** [review]

**Pull request merged**